### PR TITLE
Represent symmetrically quantized weights in signed data type

### PR DESCRIPTION
**Pull request by l-bat**
### Changes

Represent symmetrically quantized weights in signed data type with no zero point

### Reason for changes

* To detect the quantization type without analyzing zero-point values
* Signed data type for symmetrically quantized weights will lead to a smaller footprint, especially in case of grouped quantization.

### Related tickets

130625

### Tests

Updated: `tests/torch/ptq/test_weights_compression.py` and `tests/openvino/native/quantization/test_weights_compression.py`

Merge after: https://github.com/openvinotoolkit/openvino/pull/24457

Model | Backend | Metric name | Metric value | Metric diff | Num int4 | Num int8 | RAM MiB | Compr. time | Total time
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
tinyllama_data_aware_awq_scale_estimation | OV | Similarity | 0.84048 | -0.15952 | 94 | 124 | 35560 | 0:06:31 | 0:08:36
tinyllama_data_aware_awq_scale_estimation_stateful | OV | Similarity | 0.84048 | -0.15952 | 94 | 124 | 36612 | 0:06:12 | 0:07:40
tinyllama_data_aware_awq_stateful | OV | Similarity | 0.85259 | -0.14741 | 94 | 124 | 34824 | 0:01:50 | 0:03:17
tinyllama_data_aware | OV | Similarity | 0.83853 | -0.16147 | 94 | 124 | 30604 | 0:01:25 | 0:03:30
tinyllama_data_aware_gptq | OV | Similarity | 0.82187 | -0.17813 | 94 | 124 | 39624 | 0:25:09 | 0:27:10
tinyllama_data_free | OV | Similarity | 0.72057 | -0.27943 | 114 | 84 | 6671 | 0:00:42 | 0:02:46
tinyllama_int8_data_free | TORCH | Similarity | 0.95624 | -0.04376 | 0 | 312 | 30161 | 0:00:09 | 0:02:54

### CODE REVIEW

**Comment by xiao1228** [comment]
Thank you for this feature, just wondering is GPTQ model going to be automatically saved as i4 as well?
for example: `"TheBloke/Llama-2-7b-Chat-GPTQ"` which it is symmetric quantized

**Comment by l-bat** [comment]
> Thank you for this feature, just wondering is GPTQ model going to be automatically saved as i4 as well? for example: `"TheBloke/Llama-2-7b-Chat-GPTQ"` which it is symmetric quantized

No, this feature will only be enabled for weight compression via NNCF.

**Comment by xiao1228** [comment]
> > Thank you for this feature, just wondering is GPTQ model going to be automatically saved as i4 as well? for example: `"TheBloke/Llama-2-7b-Chat-GPTQ"` which it is symmetric quantized
> 
> No, this feature will only be enabled for weight compression via NNCF.

are you planning to extend it and enable this for the GPTQ model? as it will be very helpful. Current GPTQ model has a per-tensor zero point and u4 weights, which will make sense to save as i4 as symmetric.

File: tests/post_training/data/wc_reference_data.yaml [hunk 1]
```diff
@@ -1,22 +1,22 @@
 tinyllama_data_free_backend_OV:
   metric_value: 0.72057
-  num_int4: 228
+  num_int4: 114
```
Comment (ljaljushkin) [inline tests/post_training/data/wc_reference_data.yaml:3]: Why is it twice lower? I suppose, the counting of int4 layer should be modified as well

**Review by ljaljushkin — CHANGES_REQUESTED** [review]
Would be great to see the performance numbers similar to the ones from comment:
https://github.com/openvinotoolkit/nncf/pull/2537#issuecomment-2023890836

File: tests/post_training/data/wc_reference_data.yaml (hunk 1 above)
Comment (l-bat) [reply tests/post_training/data/wc_reference_data.yaml:3]: Because in the symmetric case, we don't have zero points anymore.

Comment (ljaljushkin) [reply tests/post_training/data/wc_reference_data.yaml:3]: I see, they were counted with weights as well.

**Comment by l-bat** [comment]
Develop
Model | Backend | Metric name | Metric value | Metric diff | Num int4 | Num int8 | RAM MiB | Compr. time | Total time
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
tinyllama_data_aware_awq_scale_estimation | OV | Similarity | 0.8404 | -0.1596 | 188 | 124 | 37513 | 0:05:50 | 0:07:52
tinyllama_data_aware_awq_scale_estimation_stateful | OV | Similarity | 0.8404 | -0.1596 | 188 | 124 | 39704 | 0:05:43 | 0:07:10
tinyllama_data_aware_awq_stateful | OV | Similarity | 0.85259 | -0.14741 | 188 | 124 | 29770 | 0:01:39 | 0:03:04
tinyllama_data_aware | OV | Similarity | 0.83853 | -0.16147 | 188 | 124 | 30731 | 0:01:09 | 0:03:11
tinyllama_data_free | OV | Similarity | 0.72057 | -0.27943 | 228 | 84 | 6588 | 0:00:31 | 0:02:33
tinyllama_int8_data_free | TORCH | Similarity | 0.95624 | -0.04376 | 0 | 312 | 33420 | 0:00:05 | 0:02:49



l-bat:lt/wc_sym_signed
Model | Backend | Metric name | Metric value | Metric diff | Num int4 | Num int8 | RAM MiB | Compr. time | Total time
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
tinyllama_data_aware_awq_scale_estimation | OV | Similarity | 0.84048 | -0.15952 | 94 | 124 | 37752 | 0:06:03 | 0:08:06
tinyllama_data_aware_awq_scale_estimation_stateful | OV | Similarity | 0.84048 | -0.15952 | 94 | 124 | 39638 | 0:05:52 | 0:07:19
tinyllama_data_aware_awq_stateful | OV | Similarity | 0.85259 | -0.14741 | 94 | 124 | 29995 | 0:01:47 | 0:03:13
tinyllama_data_aware | OV | Similarity | 0.83853 | -0.16147 | 94 | 124 | 30762 | 0:01:16 | 0:03:19
tinyllama_data_free | OV | Similarity | 0.72057 | -0.27943 | 114 | 84 | 6574 | 0:00:37 | 0:02:38
tinyllama_int8_data_free | TORCH | Similarity | 0.95624 | -0.04376 | 0 | 312 | 33338 | 0:00:05 | 0:02:47

**Review by ljaljushkin — APPROVED** [review]

File: nncf/quantization/algorithms/weight_compression/openvino_backend.py [hunk 2]
```diff
@@ -176,15 +175,18 @@ def transform_model(
...
-                    compressed_weight.zero_point.data,
-                    dtype=compression_dtype,
-                    name=f"{const_node_name}/zero_point",
-                )
-                converted_zero_point = opset.convert(zero_point_const, ov.Type.f16)
-                converted_const = opset.subtract(
-                    converted_const, converted_zero_point, name=f"{const_node_name}/zero_point/subtract"
-                )
+                if compressed_weight.tensor.dtype == TensorDataType.int8:
+                    assert count_nonzero(compressed_weight.zero_point.data) == 0
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/openvino_backend.py:179]: I hope tests for `compress_weight` function enough and you can delete this code.

File: nncf/quantization/algorithms/weight_compression/torch_backend.py [hunk 3]
```diff
@@ -225,13 +231,15 @@ def transform_model(
...
 
-            # pack zero point tensor
-            packed_zero_point = compressed_weight.zero_point.astype(TensorDataType.uint8)
-
             # creates weight decompressor
-            decompressor = WeightsDecompressor(
-                compressed_weight.scale.data, packed_zero_point.data, result_dtype=weight.dtype
-            )
+            if compression_config.mode == CompressWeightsMode.INT8_SYM:
+                assert count_nonzero(compressed_weight.zero_point) == 0
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/torch_backend.py:236]: The same comment.

File: nncf/torch/quantization/layers.py [hunk 4]
```diff
@@ -1064,3 +1064,24 @@ def forward(self, x):
...
+        """
+        :param scale: A scale in quantization scheme
+        :param result_dtype: (Optional) A data type that result should be cast to
+        """
+        super().__init__()
+        self.register_buffer("_scale", scale)
+        self.result_dtype = result_dtype
+
+    def forward(self, x):
+        zero_point = torch.zeros_like(self._scale)
```
Comment (alexsu52) [inline nncf/torch/quantization/layers.py:1084]: It look like a bad idea to create zero point tensor on each forward. Could you consider to introduce decompress operation for symmetric quantization?

File: nncf/quantization/algorithms/weight_compression/weight_lowering.py [hunk 5]
```diff
@@ -222,25 +222,22 @@ def calculate_integer_quantization_params(
...
     else:
-        level_low_sym = -(2 ** (num_bits - 1))
-        level_high_sym = 2 ** (num_bits - 1) - 1
-
+        level_high = 2 ** (num_bits - 1) - 1
         scale = fns.max(fns.abs(weight), axis=reduction_axes, keepdims=True)  # [a1, r//gs, 1, a2]
-        scale = scale / level_high_sym
-        zero_point = fns.as_tensor_like(scale, [-level_low_sym]).astype(TensorDataType.int32)
+        scale = scale / level_high
+        zero_point = fns.as_tensor_like(scale, [0]).astype(TensorDataType.int32)
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/weight_lowering.py:240]: I would suggest to refactor weight_lowering to avoid creation unnecessary tensors and then checking data of tensor in the weight compression transformation.

File: tests/post_training/data/wc_reference_data.yaml [hunk 6]
```diff
@@ -1,22 +1,22 @@
...
-  num_int4: 188
+  num_int4: 94
   num_int8: 124
 tinyllama_data_aware_awq_stateful_backend_OV:
   metric_value: 0.85259
-  num_int4: 188
+  num_int4: 94
   num_int8: 124
 tinyllama_data_aware_awq_scale_estimation_backend_OV:
   metric_value: 0.83795
```
Comment (ljaljushkin) [inline tests/post_training/data/wc_reference_data.yaml:14]: please use the exact metric values that you've got in local and ci run: 0.84048

File: tests/post_training/data/wc_reference_data.yaml [hunk 7]
```diff
@@ -1,22 +1,22 @@
...
-  num_int4: 188
+  num_int4: 94
   num_int8: 124
 tinyllama_data_aware_awq_scale_estimation_stateful_backend_OV:
   metric_value: 0.83795
```
Comment (ljaljushkin) [inline tests/post_training/data/wc_reference_data.yaml:18]: please use the exact metric values that you've got in local and ci run: 0.84048

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: tests/post_training/data/wc_reference_data.yaml (hunk 7 above)
Comment (l-bat) [reply tests/post_training/data/wc_reference_data.yaml]: updated

File: tests/post_training/data/wc_reference_data.yaml [hunk 8]
```diff
@@ -11,18 +11,18 @@ tinyllama_data_aware_awq_stateful_backend_OV:
...
   num_int4: 94
   num_int8: 124
 tinyllama_int8_data_free_backend_TORCH:
   metric_value: 0.95624
   num_int4: 0
   num_int8: 312
 tinyllama_data_aware_gptq_backend_OV:
-  metric_value: 0.79797
-  num_int4: 188
+  metric_value: 0.82187
```
Comment (ljaljushkin) [inline tests/post_training/data/wc_reference_data.yaml:26]: @alexsu52 is it expected improvement?

Comment (alexsu52) [reply tests/post_training/data/wc_reference_data.yaml:26]: I don't wait any improvements from this changes. I would suggest to check that scales and zero_points with/without PR.

**Comment by l-bat** [comment]
ci job: 23

File: tests/post_training/data/wc_reference_data.yaml (hunk 8 above)
Comment (l-bat) [reply tests/post_training/data/wc_reference_data.yaml]: fixed in https://github.com/openvinotoolkit/nncf/pull/2725

File: nncf/torch/quantization/quantize_functions.py [hunk 9]
```diff
@@ -259,5 +259,7 @@ def decompress(input: torch.Tensor, scale: torch.Tensor, zero_point: torch.Tenso
     :return: The decompressed tensor
     """
     input = input.type(dtype=scale.dtype)
-    decompressed_input = (input - zero_point) * scale
+    if zero_point is not None:
```
Comment (alexsu52) [inline nncf/torch/quantization/quantize_functions.py:262]: I would suggest implementing a separate function for asymmetric and symmetric cases to make tracing easier.

**Review by ljaljushkin — APPROVED** [review]

File: nncf/quantization/algorithms/weight_compression/scale_estimation.py [hunk 10]
```diff
@@ -300,3 +297,18 @@ def apply(
...
+
+def get_target_zero_mask(compressed_weights, zp=None):
+    target = compressed_weights
+    if zp is not None:
+        target = target.astype(dtype=zp.dtype) - zp
+    zero_mask = fns.isclose(target, 0)
+    return target, zero_mask
+
+
+def get_near_to_ideal_scale(weight, target, zero_mask, importance):
```
Comment (AlexKoff88) [inline nncf/quantization/algorithms/weight_compression/scale_estimation.py:310]: Can we change this funny name to something more earthly) For example, `estimate_scales`, `tune_scales`, etc.

**Review by AlexKoff88 — APPROVED** [review]

File: nncf/quantization/algorithms/weight_compression/scale_estimation.py [hunk 11]
```diff
@@ -165,8 +165,6 @@ def apply(
             original_weight = fns.zeros_like(weight) + weight
 
             compressed_weights, scale, zp = do_integer_quantization(original_weight, reduction_axis, config)
-            zp = zp.astype(scale.dtype)
```
Comment (andreyanufr) [inline nncf/quantization/algorithms/weight_compression/scale_estimation.py:168]: if zp is not None:
    zp = zp.astype(scale.dtype)
this conversion is important for performance

File: nncf/quantization/algorithms/weight_compression/scale_estimation.py [hunk 12]
```diff
@@ -300,3 +299,18 @@ def apply(
             res[weight_name] = result_scale
 
         return res
+
+
+def get_target_zero_mask(compressed_weights, zp=None):
+    target = compressed_weights
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/scale_estimation.py:305]: typing? docstrings?

Comment (l-bat) [reply nncf/quantization/algorithms/weight_compression/scale_estimation.py:313]: done

**Review by alexsu52 — APPROVED** [review]
LGTM
